File: stacked-ios/Stacked/Services/InstallmentProgress.swift
import Foundation

/// Detecta parcelas geradas (`titulo … / Parcela N`) e resume progresso + valor restante.
enum InstallmentProgress {
  static let titleMarker = " / Parcela "

  struct Snapshot: Equatable {
    let done: Int
    let total: Int
    /// Soma do `valor` das parcelas ainda não pagas. `nil` se nenhuma parcela tem valor.
    let remainingValor: Double?

    var fraction: Double {
      guard total > 0 else { return 0 }
      return Double(done) / Double(total)
    }

    var label: String {
      let count = "\(done)/\(total) pagas"
      guard let remainingValor else { return count }
      if remainingValor <= 0 {
        return done >= total ? "\(done)/\(total) pagas" : count
      }
      return "\(count) · \(CurrencyFormat.brl(remainingValor)) restante"
    }
  }

  static func isInstallmentTitle(_ title: String) -> Bool {
    title.contains(titleMarker)
  }

  /// `nil` se a tarefa não parece um parcelamento (≥2 subtarefas com o marcador).
  static func snapshot(
    from subtasks: [Subtask],
    isDone: (Subtask) -> Bool = { $0.done }
  ) -> Snapshot? {
    let parcels = subtasks.filter { isInstallmentTitle($0.title) }
    guard parcels.count >= 2 else { return nil }

    let done = parcels.reduce(into: 0) { count, sub in
      if isDone(sub) { count += 1 }
    }
    let unpaidValores = parcels.compactMap { sub -> Double? in
      guard !isDone(sub), let valor = sub.valor else { return nil }
      return valor
    }
    let anyValor = parcels.contains { $0.valor != nil }
    let remaining: Double? = anyValor ? unpaidValores.reduce(0, +) : nil

    return Snapshot(done: done, total: parcels.count, remainingValor: remaining)
  }
}

File: stacked-web/lib/utils/installment-progress.ts
/**
 * Paridade iOS InstallmentProgress — detecta `… / Parcela N` e resume progresso + valor restante.
 */

import type { Subtask } from "@/lib/types/task";
import { formatInstallmentValor } from "@/lib/utils/installment-generator";

export const INSTALLMENT_TITLE_MARKER = " / Parcela ";

export type InstallmentProgressSnapshot = {
  done: number;
  total: number;
  /** Soma do valor das parcelas ainda não pagas. null se nenhuma tem valor. */
  remainingValor: number | null;
  fraction: number;
  label: string;
};

export function isInstallmentTitle(title: string): boolean {
  return title.includes(INSTALLMENT_TITLE_MARKER);
}

function installmentLabel(
  done: number,
  total: number,
  remainingValor: number | null,
): string {
  const count = `${done}/${total} pagas`;
  if (remainingValor == null) return count;
  if (remainingValor <= 0) {
    return done >= total ? `${done}/${total} pagas` : count;
  }
  return `${count} · ${formatInstallmentValor(remainingValor)} restante`;
}

/** null se a tarefa não parece parcelamento (≥2 subtarefas com o marcador). */
export function installmentProgressSnapshot(
  subtasks: Subtask[] | undefined | null,
): InstallmentProgressSnapshot | null {
  if (!subtasks?.length) return null;
  const parcels = subtasks.filter((s) => isInstallmentTitle(s.name));
  if (parcels.length < 2) return null;

  let done = 0;
  const unpaidValores: number[] = [];
  let anyValor = false;

  for (const sub of parcels) {
    if (sub.done) done += 1;
    if (sub.valor != null && Number.isFinite(sub.valor)) {
      anyValor = true;
      if (!sub.done) unpaidValores.push(sub.valor);
    }
  }

  const remainingValor = anyValor
    ? unpaidValores.reduce((sum, v) => sum + v, 0)
    : null;
  const total = parcels.length;
  const fraction = total > 0 ? done / total : 0;

  return {
    done,
    total,
    remainingValor,
    fraction,
    label: installmentLabel(done, total, remainingValor),
  };
}
